src/rt/sched/behaviourcore.h

```diff
@@ -735,178 +740,226 @@ namespace verona::rt
             continue;
           }
 
           yield();
           bool first_reader = cown->read_ref_count.add_read();
-          Logging::cout() << " Reader got the cown " << *curr_slot
+          Logging::cout() << " Reader got the cown " << *new_slot
                           << Logging::endl;
           yield();
-          curr_slot->set_read_available();
-          ec[std::get<0>(indexes[first_chain_index])]++;
+          new_slot->set_read_available();
+          ec[first_body_index]++;
           if (first_reader)
           {
             Logging::cout()
               << "Acquiring reference count for first reader on cown "
-              << *curr_slot << Logging::endl;
+              << *new_slot << Logging::endl;
             Cown::acquire(cown);
           }
           continue;
         }
 
         Logging::cout()
           << " Writer waiting for cown. Set next of previous slot cown "
-          << *curr_slot << " previous " << *prev_slot << Logging::endl;
+          << *new_slot << " previous " << *prev_slot << Logging::endl;
         prev_slot->set_next_slot_writer(first_body);
         yield();
       }
 
-      // Second phase - Release phase.
+      // Third phase - Release phase.
       for (size_t i = 0; i < body_count; i++)
       {
         Logging::cout() << "Release phase for behaviour " << bodies[i]
                         << Logging::endl;
       }
-      for (size_t i = 0; i < count; i++)
+      // FIXME: Shouldn't this be only over chains?
+      // Do we avoid the if this way?
+      for (size_t i = 0; i < cown_count; i++)
       {
         yield();
-        auto slot = std::get<1>(indexes[i]);
+        auto slot = std::get<1>(cown_to_behaviour_slot_map[i]);
         Logging::cout() << "Setting slot " << slot << " to ready"
                         << Logging::endl;
         if (slot->is_wait_2pl())
           slot->set_ready();
       }
 
+      // Fourth phase - Process & Resolve
+
+      // Process first reference count and the chains with no predecessor
+      for (size_t i = 0; i < chain_count; i++)
+      {
+        auto* cown = std::get<0>(chain_info[i]);
+        auto first_body_index = std::get<1>(chain_info[i]);
+        auto* first_body = bodies[first_body_index];
+        auto* curr_slot = std::get<2>(chain_info[i]);
+        auto chain_had_no_predecessor = std::get<4>(chain_info[i]);
+        auto transfer_count = std::get<3>(chain_info[i]);
+
+        if (chain_had_no_predecessor)
+        {
+          if (curr_slot->is_read_only())
+          {
+            yield();
+            bool first_reader = cown->read_ref_count.add_read();
+            Logging::cout() << "Reader at head of queue and got the cown "
+                            << *curr_slot << Logging::endl;
+            yield();
+            curr_slot->set_read_available();
+            ec[first_body_index]++;
+            yield();
+            acquire_with_transfer(cown, transfer_count, 1 + first_reader);
+            continue;
```

Comment:
When we make the multischedule work with readonly, I think this part of the code will look closer to release.